linera-core/src/chain_worker/delivery_notifier.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module contains a helper type to notify when cross-chain messages have been
+//! delivered.
+//!
+//! Keeping it in a separate module ensures that only the chain worker is able to call its
+//! methods.
+
+use std::{
+    collections::BTreeMap,
+    mem,
+    sync::{Arc, Mutex},
+};
+
+use linera_base::data_types::BlockHeight;
+use tokio::sync::oneshot;
+use tracing::warn;
+
+/// A set of pending listeners waiting to be notified about the delivery of messages sent
+/// from specific [`BlockHeight`]s.
+///
+/// The notifier instance can be cheaply `clone`d and works as a shared reference.
+/// However, its methods still require `&mut self` to ensure that it is only changed by
+/// [`ChainWorkerStateWithAttemptedChanges`][`super::ChainWorkerStateWithAttemptedChanges`].
+#[derive(Clone, Default)]
+pub struct DeliveryNotifier {
+    notifiers: Arc<Mutex<BTreeMap<BlockHeight, Vec<oneshot::Sender<()>>>>>,
+}
+
+impl DeliveryNotifier {
+    /// Returns [`true`] if there are no pending listeners.
```

Comment:
Does that work… and is it useful? :smile:

```suggestion
    /// Returns `true` if there are no pending listeners.
```